daemon/cluster/cluster.go

```diff
@@ -625,7 +605,16 @@ func (c *Cluster) Update(version uint64, spec types.Spec, flags types.UpdateFlag
 		return err
 	}
 
-	swarmSpec, err := convert.SwarmSpecToGRPC(spec)
+	// In update, client should specify the desired state, which includes
+	// Name and Labels. So we need to initialize the name and labels here
+	// based on the value in spec.
+	// It is possible that empty Labels be specified (unset in update).
+	swarmSpec, err := convert.SwarmSpecToGRPC(spec, swarmapi.ClusterSpec{
+		Annotations: swarmapi.Annotations{
+			Name:   spec.Name,
+			Labels: spec.Labels,
+		},
+	})
```

Comment:
Thanks @aaronlehmann. At the client side (`api/client/swarm/update.go`) the existing swarm setting was obtained and then merged with the setting updates. So the updated setting is reflected in `spec` here (without `getSwarm`).

But it is better to  obtain the existing swarm setting in daemon side here (with `getSwarm` and then update as needed) as end user may not use client directly (vs. remote API). Processing at the daemon side will prevent incorrect settings through remote API.

Let me update the pull request to change it.